astropy/units/tests/test_quantity_non_ufuncs.py

```diff
@@ -44,42 +46,54 @@ def allows_array_function_override(f):
     else:
         from numpy.testing.overrides import allows_array_function_override
 
-    wrapped_functions = {}
+    wrapped_functions = {mod.__name__: {} for mod in modules}
     for mod in modules:
         for name, f in mod.__dict__.items():
             if callable(f) and allows_array_function_override(f):
-                # Indexing by just the name is easiest for test writing,
-                # but in numpy 2.0, there are 2 versions of diagonal and
-                # trace, one in the main namespace, and one in np.linalg.
-                # So, we distinguish those by adding the module name.
-                if name in wrapped_functions:
-                    name = mod.__name__.replace("numpy.", "") + "_" + name
-                wrapped_functions[name] = f
+                wrapped_functions[mod.__name__][name] = f
 
     return wrapped_functions
 
 
 all_wrapped_functions = get_wrapped_functions(
     np, np.fft, np.linalg, np.lib.recfunctions
 )
-all_wrapped = set(all_wrapped_functions.values())
+
+all_wrapped = set(
+    itertools.chain.from_iterable(_.values() for _ in all_wrapped_functions.values())
+)
+
+MOD_REGEXP = re.compile(r"((?P<module_name>[A-Z_]+)_)?(?P<function_name>[\w_]+)")
+
+
+def populate_covered_from_members(members, covered: set[FunctionType]) -> None:
+    for k, v in members.items():
+        if inspect.isfunction(v) and k.startswith("test_"):
+            if (m := MOD_REGEXP.fullmatch(k.removeprefix("test_"))) is None:
+                raise RuntimeError(f"failed to match {k=}")
+            function_name = m.group("function_name")
+            if (module_name := m.group("module_name")) is not None:
+                module_name = "numpy." + module_name.replace("_", ".").lower()
+            else:
+                module_name = "numpy"
+
+            if (
+                func := all_wrapped_functions[module_name].get(function_name)
+            ) is not None:
+                covered.add(func)
 
 
 class CoverageMeta(type):
     """Meta class that tracks which functions are covered by tests.
 
-    Assumes that a test is called 'test_<function_name>'.
+    Assumes that a test is called 'test_<MODULE_NAME>_<function_name>',
+    where MODULE_NAME (and its trailing '_') may be missing.
     """
 
     covered = set()
 
     def __new__(mcls, name, bases, members):
```

Comment:
Thought: if we change this to `__init_subclass__`, then we could have the signature of a test class be,
```
class TestFFT(InvariantUnitTestSetup, tested_module="np.fft"):
    ...
```

Reply:
This seems like a good idea, however I feel slightly out of my depth here, and I haven't succeeded in implementing it yet.
Could you show me how to do this change in a little bit more details, or maybe point me to relevant docs that I could use to learn about it ?